File: 2018/src/day05.rs
use std::fs;

fn react(polymer: &str, skip: char) -> String {
    let (mut reacted, mut last) = (Vec::new(), ' ');
    for char in polymer.chars() {
        if char.eq_ignore_ascii_case(&skip) {
            continue;
        }
        if char.eq_ignore_ascii_case(&last) && char != last {
            reacted.pop();
            last = match reacted.last() {
                Some(&c) => c,
                None => ' ',
            };
            continue;
        }
        reacted.push(char);
        last = char;
    }
    reacted.into_iter().collect::<String>()
}

pub fn day05() {
    let input = "data/input-05.txt";
    let polymer = fs::read_to_string(input).unwrap().replace('\n', "");

    let res = react(&polymer, ' ').len();
    println!("Part 1: {res}");
    assert_eq!(res, 9296);

    let res = ('a'..='z').fold(polymer.len(), |min, c| {
        usize::min(min, react(&polymer, c).len())
    });
    println!("Part 2: {res}");
    assert_eq!(res, 5534);
}

File: 2018/nim/day05.nim
import strutils, sugar

proc react(polymer: string, skip: char): string =
    var reacted: seq[char]
    var last = ' '
    for ch in polymer:
        if ch.toLowerAscii() == skip:
            continue
        if ch != last and ch.toLowerAscii() == last.toLowerAscii():
            reacted.delete(len(reacted))
            last = if len(reacted) > 0: reacted[len(reacted) - 1] else: ' '
            continue
        reacted.add(ch)
        last = ch
    return reacted.join("")

let polymer = readFile("data/input-05.txt").strip().splitLines()[0]

let res1 = len(react(polymer, ' '))
echo "Part 1: ", res1
assert(res1 == 9296)

let all = collect(newSeq):
    for c in 'a'..'z':
        len(react(polymer, c))
let res2 = min(all)
echo "Part 2: ", res2
assert(res2 == 5534)
